controllers/purge_controller.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	controllerutil.AddFinalizer(kyma, v1beta1.PurgeFinalizer)
+	if err := r.Update(ctx, kyma); err != nil {
+		r.Event(kyma, "Warning", "SettingPurgeFinalizerError", fmt.Errorf("could not set purge finalizer: %w", err).Error())
+		return err
+	}
+	return nil
+}
+
+func (r *PurgeReconciler) DropPurgeFinalizer(ctx context.Context, kyma *v1beta1.Kyma) error {
+	if controllerutil.ContainsFinalizer(kyma, v1beta1.PurgeFinalizer) {
+		controllerutil.RemoveFinalizer(kyma, v1beta1.PurgeFinalizer)
+		if err := r.Update(ctx, kyma); err != nil {
+			r.Event(kyma, "Warning", "SettingPurgeFinalizerError",
+				fmt.Errorf("could not remove purge finalizer: %w", err).Error())
+			return err
+		}
+	}
+
+	return nil
+}
+
+func performCleanup(ctx context.Context, remoteClient client.Client, skipCRDs CRDMatcher) error {
+	crdList := apiextensions.CustomResourceDefinitionList{}
+
+	if err := remoteClient.List(ctx, &crdList); err != nil {
+		return err
+	}
+
+	for _, crdResource := range crdList.Items {
+		if isKymaCR(crdResource) {
+			continue
+		}
+
+		if skipCRDs(crdResource) {
+			continue
+		}
+
+		staleResources, err := getStaleResourcesFrom(ctx, remoteClient, crdResource)
+		if err != nil {
+			return err
+		}
+
+		err = purgeStaleResources(ctx, remoteClient, staleResources)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getStaleResourcesFrom(ctx context.Context, remoteClient client.Client,
+	crd apiextensions.CustomResourceDefinition,
+) (unstructured.UnstructuredList, error) {
+	staleResources := unstructured.UnstructuredList{}
+	// Since there are multiple possible versions, we are choosing the one that's in the etcd storage
+	var gvkVersion string
+	for _, version := range crd.Spec.Versions {
+		if version.Storage {
+			gvkVersion = version.Name
+			break
+		}
+	}
+
+	gvk := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Kind:    crd.Spec.Names.Kind,
+		Version: gvkVersion,
+	}
+
+	staleResources.SetGroupVersionKind(gvk)
+
+	if err := remoteClient.List(ctx, &staleResources); err != nil {
+		return unstructured.UnstructuredList{}, err
+	}
+
+	return staleResources, nil
+}
+
+func purgeStaleResources(ctx context.Context, remoteClient client.Client,
+	staleResources unstructured.UnstructuredList,
+) error {
+	for index := range staleResources.Items {
+		resource := staleResources.Items[index]
+		resource.SetFinalizers(nil)
+		if err := remoteClient.Update(ctx, &resource); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *PurgeReconciler) UpdateStatus(
+	ctx context.Context, kyma *v1beta1.Kyma, state v1beta1.State, message string,
+) error {
+	if err := status.Helper(r).UpdateStatusForExistingModules(ctx, kyma, state, message); err != nil {
+		return fmt.Errorf("error while updating status to %s because of %s: %w", state, message, err)
+	}
+	return nil
+}
+
+func (r *PurgeReconciler) RecordKymaStatusMetrics(_ context.Context, _ *v1beta1.Kyma) {}
+
+func (r *PurgeReconciler) IsKymaManaged() bool { return false }
```

Comment:
Why is this always false? You could also use the flag bool, as in the kyma controller